drv/i2c-devices/src/emc2305.rs

```diff
@@ -0,0 +1,334 @@
+            return Err(ResponseCode::BadArg);
+        }
+
+        assert!(fan_count <= MAX_FANS);
+        let device = &self.device;
+
+        // Enable the watchdog at all times
+        let mut config =
+            Configuration(read_reg8(device, Register::Configuration)?);
+        config.set_watchdog_enable(true);
+        write_reg8(device, Register::Configuration, config.0)?;
+
+        for fan in 0..fan_count {
+            let fan = Fan::try_from(fan).unwrap();
+
+            // Configure tach stuff
+            let reg1 = fan.configuration1();
+            let mut cfg1 = FanCfg1(read_reg8(device, reg1)?);
+            cfg1.set_range(0b01); // 1000 RPM minimum, TACH count multiple = 2
+            cfg1.set_edges(0b01); // 5 edges sampled, TACH multiple = 1x
+            write_reg8(device, reg1, cfg1.0)?;
+
+            // Enable ramp control, to avoid the fans going from 0-100%
+            let reg2 = fan.configuration2();
+            let mut cfg2 = FanCfg2(read_reg8(device, reg2)?);
+            cfg2.set_enable_ramp_control(true);
+            write_reg8(device, reg2, cfg2.0)?;
+
+            write_reg8(device, fan.pwm_target(), 0)?;
+        }
+
+        Ok(())
+    }
+
+    /// Determines the rotations per minute based on the tach count
+    pub fn fan_rpm(&self, fan: Fan) -> Result<Rpm, ResponseCode> {
+        let val = read_reg16(&self.device, fan.tach_count())?;
+
+        // See section 4.4.1 in the datasheet for details
+        //
+        // The tachometer reading represents the number of 32 KHz clock cycles
+        // counted in a single revolution of the fan.
+        //
+        // So to get from the tach count to the time per revolution:
+        //
+        //              1         n - 1
+        //      RPM = ----- x ------------- x f_TACH x 60
+        //            poles    count * 1/m
+        //
+        // Our settings are the following:
+        //   poles = 2
+        //   n = edges measured (set to 5 in `cfg1` above)
+        //   m = range multiplier (2x, per configuration above)
+        //   f_TACH = tachometer measurement frequency (32 KHz)
+        //   count = reading
+        //
+        // Doing the math, we end up with
+        //   RPM = 7680000 / count
+        let count = ((val[0] as u32) << 5) | (val[1] >> 3) as u32;
+
+        if count == 0 {
+            // What happens if the fan is idle / stalled / unplugged?
+            //
+            // On the MAX31790, it returns all ones for the tach reading, but
+            // it's not clear whether that's also the case for this chip.
+            ringbuf_entry!(Trace::ZeroTach(fan));
+            Err(ResponseCode::BadDeviceState)
+        } else {
+            let rpm = 7_680_000 / count;
+            let Ok(rpm) = rpm.try_into() else {
+                ringbuf_entry!(Trace::TachOverflow(count));
+                return Err(ResponseCode::BadDeviceState);
+            };
+            Ok(Rpm(rpm))
+        }
```

Comment:
very nitpicky, but: my preference would be to write the error condition as an early return, as:
```suggestion
            return Err(ResponseCode::BadDeviceState);
        }
        
        let rpm = 7_680_000 / count;
        let Ok(rpm) = rpm.try_into() else {
            ringbuf_entry!(Trace::TachOverflow(count));
            return Err(ResponseCode::BadDeviceState);
        };
        Ok(Rpm(rpm))
```